ENH: rework global marker handling

ZLLentz opened the pull request:
See https://jira.slac.stanford.edu/browse/ECS-7128 for the full motivation and context.

### Main Goals
- Make "global" marker mode use all four markers instead of only the first two markers.
- Allow seamless transitions between local and global marker modes.
- Make it really clear which markers are global and which are local via marking them with G or L.
- Completely separate the saved local marker positions from the global marker positions, so that neither set overrides the other.
- Allow for "partial" loading of global markers in the event that only some of the PVs exist.

#### Notable Internal Fixes / Adjustments
- Allow us to cleanly swap markers in and out in DisplayImage to help separate global and local marker position data in memory.
- Remove vestigial code for an older version that had weird special global 2 mode for markers 3, 4 with different behavior.
- Improve the performance of the global marker code by removing an unnecessary dumpConfig step, which suddenly was slowing down the GUI when I doubled the number of global markers.
- Rework the config loading code to remove strange code branches and give us consistent first-load behavior for new cameras.

![image](https://github.com/user-attachments/assets/f4446530-beeb-41fd-8f2d-fadf8c5cf8a6)

ZLLentz committed:
ENH: rework global marker handling

Include markers 3, 4 in the global marker pool if they connect.
Indicate clearly which markers are global and which are local,
this is helpful if the IOC doesn't have all 4.
Improve performance of marker code, doesn't slow down the GUI as much.
Remove vestigial code for an older version that had weird special
global 2 mode for markers 3, 4 with different behavior.

ZLLentz commented:
Clicking around a bit, the first issue I see is that my new button doesn't properly show the status when switching to a new camera. Oops!

The per-marker indicator is always correct because it's triggered specifically by the PV callback routines.

ZLLentz committed:
FIX: ensure side buttons are synchronized too

ZLLentz committed:
FIX: make sure to reset the marker PVs while cleaning up a cam

ZLLentz commented:
One more TODO per conversation in the Jira: I will stop having the global marker positions clobber the local marker positions. These two position sets will be completely separate.

ZLLentz committed:
ENH: progress, global and local marker points are almost fully separated.

ZLLentz committed:
ENH: properly re-enable partial global markers for missing PV cases

ZLLentz committed:
FIX: clarify that reset only applies to local markers now

ZLLentz committed:
FIX: don't error if markers are configured before first camera

ZLLentz committed:
STY: run isort once because the import order was annoying me

ZLLentz commented:
This works essentially flawlessly now, except for on cameras you've never loaded before. I'll round out my day by tracking down why this fails when there isn't a config file to load.

ZLLentz committed:
ENH: rework config loading so it works properly with a new cam or missing fields

ZLLentz committed:
DOC: add docstrings to added functions in DisplayImage

ZLLentz committed:
MNT: remove unused function I added

ZLLentz committed:
DOC: update some of the docstrings with new args

janeliu-slac commented on DisplayImage.py:
```diff
@@ -466,3 +456,66 @@ def roiSet(self, x, y, w, h, rel=False):
+    vertically away from each of the corners of the image, not on the image
+    itself but off of it.
+    """
+    markers[0].setAbs(-100, 100)
```
For `markers[0].setAbs(-100, 100)` should the second value be `-100`?

ZLLentz replied:
This depends on the coordinate system used and since this app does a lot of stuff internally I have no idea what the actual coordinate system is. I have the same intuition as you that this should be (-100, -100), I tried to copy what was implemented before I got here and didn't think too much about it. I can double-check that resetting the markers puts them in correct-looking positions.

KaushikMalapati replied:
I also think it should be -100, -100 initially
![image](https://github.com/user-attachments/assets/3fbbb9b8-e293-4807-a6d4-43024b8705ee)

ZLLentz replied:
I'll fix these, thanks for pointing it out and confirming

ZLLentz replied:
ok yes I also see now that I made a copy/paste error somehow when putting this all in one place, whoops!

janeliu-slac commented on DisplayImage.py:
```diff
@@ -466,3 +456,66 @@ def roiSet(self, x, y, w, h, rel=False):
+    vertically away from each of the corners of the image, not on the image
+    itself but off of it.
+    """
+    pts = [param.Point(0, 0), param.Point(0, 0), param.Point(0, 0), param.Point(0, 0)]
```
I'm not too familiar with PyQt. Is it necessary to set pts to `(0, 0)` values before calling `reset_markers(pts)`?

ZLLentz replied:
I think I just wanted to allocate "any" point and the API required an x and a y. `param.Point` isn't actually a qt thing, it's something they wrote into the param.py file in this repo. I could have put in the default values again but I didn't want to implement the defaults in multiple places.

janeliu-slac commented on camviewer_ui_impl.py:
```diff
@@ -886,41 +877,64 @@ def onSetROI2(self):
-            self.updateMarkerText(False, True, 1 << n, 1 << n)
-        else:
-            self.updateMarkerText(False, True, 0, 1 << n)
+        self.updateMarkerText(False, True, 1 << n, 1 << n)
```
Out of curiosity, why is it shifting the bits left?

ZLLentz replied:
Before I made any changes, there was a lot of this annoying bitshift stuff in here and I opted not to refactor it despite it being confusing.

In this case, the `updateMarkerText` method is expecting a bitmask to decide which markers to update, so the caller is bitshifting the number 1 into the right bit slot to only update one marker. There are definitely better ways to do this but I left it as it was prior to the PR.

KaushikMalapati commented on camviewer_ui_impl.py:
```diff
@@ -1865,108 +1868,40 @@ def cross2mon(self, exception=None):
                 pass
-        self.globmarkpvs2 = []
+        self.globmarkpvs = {}
+        self.ui.cross1_status.setText("L")
```
What do these do? Are they supposed to be G when you're in global mode? Ui sets them to L initially, disconnectMarkerPvs sets them to L, but they are never set to G anywhere - presumably they should be in connectMarkerPvs?

slactjohnson replied:
They appear to be set to "G" at line 1846 in onCrossUpdate.

ZLLentz replied:
They are set when we get a new value from the cross position PVs. I leave them as "L" until then in case the cross PVs never connect, so that "G" can only be reached after we've definitely connected to a real PV.

KaushikMalapati commented:
I was trying this on CRIX01 and was swapping between local and global markers in the taskbar when my camviewer became completely uninteractive. When I tried reopening that cam on prod camviewer, it said the pvs were disconnected until I restarted the ioc. I tried this branch on CRIX01 again, got a bunch of disconnect messages and then I had to restart the ioc again.

KaushikMalapati commented on camviewer_ui_impl.py:
```diff
@@ -314,10 +305,8 @@ def __init__(
-        self.useglobmarks2 = False
-        self.globmarkpvs = []
-        self.globmarkpvs2 = []
+        self.useglobmarks = True
```
If you use global markers by default I think the cross#_status widgets in the ui file should be G by default

ZLLentz replied:
This is intentional so that they don't turn to G until we successfully connect to and read from the global marker position PVs

KaushikMalapati commented on camviewer_ui_impl.py:
```diff
@@ -392,6 +381,8 @@ def __init__(
         ]
         self.markerdialog.hide()
+        self.local_marker_points = default_markers()
+        self.global_marker_points = default_markers()
```
What is global_marker_points meant to do? I see it initialized to default_markers and reset in getConfig which make sense, but I don't ever see it set to anything. new_point is set to one of its elements in onCrossUpdate, but won't it always be one of the default points?

ZLLentz replied:
I swap each global marker point into the live marker points list in display_image when it's time to fill the marker positions with data from the live PVs. This protects local_marker_points from getting filled with global PV values.

KaushikMalapati commented on camviewer_ui_impl.py:
```diff
@@ -1865,108 +1868,40 @@ def cross2mon(self, exception=None):
                 pass
-        self.globmarkpvs2 = []
+        self.globmarkpvs = {}
+        self.ui.cross1_status.setText("L")
+        self.ui.cross2_status.setText("L")
+        self.ui.cross3_status.setText("L")
+        self.ui.cross4_status.setText("L")
```
I don't think this is correct because you call disconnectMarkerPvs in connectMarkerPvs to update what they are. This should only set the text to L if useglobmarks

ZLLentz replied:
I think this is always appropriate. I need to guarantee that these say "G" if and only if their value reflects a live PV value, which is never guaranteed until we resolve the PV connection and get the first value.

ZLLentz commented:
> swapping between local and global markers in the taskbar when my camviewer became completely uninteractive

This one is concerning to me. I noticed elsewhere that swapping camera really quickly caused freezing behavior, does the same happen when swapping between global and local triggers really quickly?

KaushikMalapati commented:
> > swapping between local and global markers in the taskbar when my camviewer became completely uninteractive
> 
> This one is concerning to me. I noticed elsewhere that swapping camera really quickly caused freezing behavior, does the same happen when swapping between global and local triggers really quickly?

Yes

ZLLentz commented:
I'll try to track down which thing I'm doing causes the bad behavior- maybe disconnecting a pyca PV when it is initializing breaks things, for example? It'd be good to document specifically where the deadlock happens.

Probably I'll have to end up (temporarily) disabling certain UI elements after the user changes them.

ZLLentz commented:
This is how I found (one of) the freezes and identified which line of code we were stuck on:
- start GUI
- select cam via combobox
- mash up and down arrows until it freezes
- py-spy
```
(pcds-5.9.1)zlentz@psbuild-rhel7-01:~$ py-spy dump --pid 6395
Process 6395: python ./camviewer.pyw --instrument xcs --pvlist /cds/group/pcds/pyps/config/xcs/camviewer.cfg
Python v3.9.7 (/cds/group/pcds/pyps/conda/py39/envs/pcds-5.1.1/bin/python3.9)

Thread 6395 (idle): "MainThread"
    disconnect (psp/Pv.py:429)
    disconnectPv (camviewer_ui_impl.py:1784)
    connectCamera (camviewer_ui_impl.py:1911)
    onCameraSelect (camviewer_ui_impl.py:2153)
    <module> (camviewer.pyw:111)
Thread 6418 (idle): "Dummy-16"
```
https://github.com/slaclab/pyca/blob/395124deff292c694ae85e1d040375b6139bfa6e/psp/Pv.py#L429
https://github.com/slaclab/pyca/blob/395124deff292c694ae85e1d040375b6139bfa6e/pyca/pyca.cc#L45

praise `py-spy` for always being there to save me

it looks like the python threads are just chilling here waiting for pyca's `clear_channel` to complete
maybe it's still not obvious how to avoid this other than forcing the user to slow down

there could be "something" to check when we disconnect Pv objects to know to wait a moment

ZLLentz committed:
FIX: default marker0 position typo

ZLLentz commented:
For continuity, here's the py-spy dump from the global marks spam method, which I couldn't get to happen on psbuild but definitely happens for CRIX01 from rix-control
```
$ py-spy dump --pid 11118
Process 11118: python ./camviewer.pyw --instrument rix --pvlist /cds/group/pcds/pyps/config/rix/camviewer.cfg
Python v3.9.7 (/cds/group/pcds/pyps/conda/py39/envs/pcds-5.1.1/bin/python3.9)

Thread 11118 (idle): "MainThread"
    disconnect (psp/Pv.py:429)
    disconnectMarkerPVs (camviewer_ui_impl.py:1896)
    setUseGlobalMarkers (camviewer_ui_impl.py:928)
    onGlobMarks (camviewer_ui_impl.py:906)
    local_clicked (camviewer_ui_impl.py:888)
    <module> (camviewer.pyw:111)
Thread 11434 (idle): "Dummy-8"
```
Very similar

In offline testing, the only thing I've been able to identify is that calling `disconnect` on a `psp.Pv` object with a monitor callback running waits until the monitor callback finishes. Possibly this is the source of the deadlock for specific monitor callbacks.

ZLLentz commented:
One of my other windows froze on close at a different part:
```
$ py-spy dump --pid 8437
Process 8437: python ./camviewer.pyw --instrument rix --pvlist /cds/group/pcds/pyps/config/rix/camviewer.cfg
Python v3.9.7 (/cds/group/pcds/pyps/conda/py39/envs/pcds-5.1.1/bin/python3.9)

Thread 8437 (idle): "MainThread"
    unsubscribe (psp/Pv.py:506)
    monitor_stop (psp/Pv.py:883)
    end_monitors (camviewer_ui_impl.py:754)
    closeEvent (camviewer_ui_impl.py:726)
    <module> (camviewer.pyw:111)
Thread 8480 (idle): "Dummy-2"
```

ZLLentz commented:
I was not super happy with any of the options I tried, so I settled on one that's at least simple:

- When the user does an action that might freeze the GUI if done multiple times quickly, disable the associated widgets for 1 second

In the process of getting to this point I tidied up much of the PV setup and cleanup code here, which all are probably good improvements but were red herrings as far as solving the base issue here.

There were multiple flavors of GUI freezing (e.g. getting stuck on a qt signal emit, getting stuck on literally nothing (?), getting stuck on app cleanup) and I was unable to figure out how to wrangle all of (any of?) them.

ZLLentz committed:
ENH/REF: prevent freezes, tidy setup and cleanup

- Force widgets disabled for 1 second after certain changes to
  prevent repeat usages of the widget.
- Ensure my monitor callbacks only register once instead of twice
  or more.
- More thoroughly and consistently clean up gui PVs.

tangkong commented on camviewer_ui_impl.py:
```diff
@@ -1200,35 +1213,20 @@ def onComboBoxColorIndexChanged(self, iNewIndex):
-            except Exception:
-                pass
-            self.putlensPv = None
+        self.camera = self.disconnectPv(self.camera)
```
To clarify how this works, we: 
* set these attributes to instances of a `psp.Pv.Pv` on startup (self.connectCamera)
* do normal things
* set these to `None` via `self.disconnectPv` as a way of deallocating the resources.

I thought it was a bit opaque, but didn't take too long to backtrack.  I think maybe a stray comment might have helped, but I think it's not a huge deal

ZLLentz replied:
You've got it right. To be clear, I'm not a big fan of this pattern that's repeated dozens of times up and down the application but I also wasn't interested in doing the work to replace it with something else.

tangkong commented on camviewer_ui_impl.py:
```diff
@@ -2415,6 +2336,9 @@ def buttonWriteCallback(self, button):
     def setupSpecific(self):
         self.ui.actionGlobalMarkers.setChecked(self.useglobmarks)
+        self.ui.global_button.setChecked(self.useglobmarks)
+        self.ui.local_button.setChecked(not self.useglobmarks)
```
Nit/Rabbithole: Apparently the way to give multiple radio buttons exclusive behaviour is to group them in a [`QButtonGroup`](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QButtonGroup.html)

I could have sworn there was an easier way to do this but I guess not.

ZLLentz replied:
The qt docs said that I could just put them in the same layout but the exclusivity didn't seem to work specifically in the case where the user changed global mode from the top menu.

ZLLentz replied:
To clarify: clicking the button reflected exclusivity, but updating the checked state myself did not.

tangkong commented on camviewer_ui_impl.py:
```diff
@@ -3067,40 +2956,70 @@ def getConfig(self):
-        self.advdialog.ui.viewWidth.setText(newwidth)
-        self.advdialog.ui.viewHeight.setText(newheight)
-        self.advdialog.ui.projSize.setText(newproj)
+        # viewwidth may be missing if there was no camera config
```
There are a lot of instances here of catching missing configs.  Perhaps the proper way to handle this is to bundle a default config file and refer to it whenever an external one is not provided.  That way we can be certain that a value exists and we can manage those values easily (rather than setting them in code?)

That being said we were supposed to stop doing camviewer work ages ago.

ZLLentz replied:
Yeah, that's a lot cleaner. I'll keep this idea in mind in general for config-reading applications in the future, if not for this one.

tangkong commented on camviewer_ui_impl.py:
```diff
@@ -3119,90 +3041,113 @@ def getConfig(self):
+            colorscale = ["Linear", "Scale"]
+        finally:
+            # Backcompat for really really really old names
+            if colorscale[0] == "Log":
```
What do people have against `Ln` for natrual log? 🤔

tangkong commented on camviewer_ui_impl.py:
```diff
@@ -886,41 +879,75 @@ def onSetROI2(self):
+        self.ui.global_button.setEnabled(allow)
+        self.ui.local_button.setEnabled(allow)
+        if not allow:
+            self.glob_changeable_restore.singleShot(1000, self.allow_glob_changes)
```
I supposed the alternative is to do some lock acquisition and disabling of the UI whenever a lengthy action is initiated?  That's a bit more involved of a refactor though.  

Stray thought: a busy cursor would probably be appreciated here if we start to deal with many lengthy actions

ZLLentz replied:
There aren't any lengthy actions here, rather there's an annoying race condition that comes from pyca:

- `disconnect` and `unsubscribe` both block and wait for any ongoing subscription callbacks to complete
- Something in the related cleanup causes the subscription callback to block
- Both threads block forever

I think it's a bilateral deadlock. It seems to have a few flavors, sometimes a signal.emit call hangs forever, for example.

tangkong replied:
Hm, troubling.  I can only assume there's some more involved way to unravel the chain of execution and ensure everything waits for its predecessors, but that's a task for later

ZLLentz replied:
Yeah I couldn't figure it out after trying for an hour or two, so I went for this plan B "Avoid the problem"

tangkong reviewed (approved):
Code looks good to me on the whole.  I think I only have nitpicks and shower thoughts to contribute.  

I took it for a spin and couldn't get it to crash/lock up in the old or new ways
